app/grandchallenge/reader_studies/forms.py

```diff
@@ -478,41 +480,174 @@ def clean_ground_truth(self):
         super().__init__(*args, **kwargs)
-        if instance is not None:
-            for (
-                slug,
-                civ,
-            ) in instance.reader_study.values_for_interfaces.items():
-                val = instance.values.filter(interface__slug=slug).first()
-
-                if civ["kind"] in InterfaceKind.interface_type_json():
-                    # Use the field/widget provided by InterfaceFormField,
-                    # which includes proper validation
-                    self.fields[slug] = InterfaceFormField(
-                        kind=civ["kind"],
-                        schema=civ["schema"],
-                        initial=val.value if val else None,
-                        required=False,
-                    ).field
-                else:
-                    # Use a ModelChoiceField here, as InterfaceFormField would
-                    # provide an upload wodget, but we do not want to add new
-                    # images/files here, but rather assign existing values to
-                    # the proper display sets.
-                    self.fields[slug] = ModelChoiceField(
-                        queryset=ComponentInterfaceValue.objects.filter(
-                            id__in=civ["values"]
-                        ),
-                        initial=val,
-                        required=False,
-                    )
-
-            self.fields["order"] = IntegerField(initial=instance.order)
+        self.instance = instance
+        self.reader_study = reader_study
+        self.user = user
+        for slug, values in reader_study.values_for_interfaces.items():
+            current_value = None
+            if instance:
+                current_value = instance.values.filter(
+                    interface__slug=slug
+                ).first()
+            interface = ComponentInterface.objects.get(slug=slug)
+            if interface.is_image_kind:
+                self.fields[slug] = self._get_image_field(
+                    interface, values, current_value
+                )
+            elif interface.requires_file:
+                self.fields[slug] = self._get_file_field(
+                    interface, values, current_value
+                )
+            else:
+                self.fields[slug] = self._get_json_field(
+                    interface, current_value
+                )
+
+        order = (
+            instance.order if instance else reader_study.next_display_set_order
+        )
+        self.fields["order"] = IntegerField(initial=order)
+
+    def _get_default_field(self, interface, current_value):
+        return InterfaceFormField(
+            instance=interface,
+            initial=current_value.value if current_value else None,
+            required=False,
+            user=self.user,
+        ).field
+
+    def _get_json_field(self, interface, current_value):
+        return self._get_default_field(interface, current_value)
```

Comment:
Why this intermediary method?

Reply:
I changed the form and methods a little, which I hope clarifies things a bit and makes a bit more sense as well. Basically for the create view I think we want to go straight to uploading new files, whereas the update view has the option to select an existing civ with a link to upload a new file.